Core/include/Acts/Propagator/GuidedNavigator.hpp

```diff
@@ -0,0 +1,230 @@
+  // See Propagator.hpp:247
+  GuidedNavigator() { throw_assert(false, "Don't use this constructor!"); }
+
+  /// Constructor of the ML Navigator
+  GuidedNavigator(Guider &&guider,
+                  std::shared_ptr<const Acts::TrackingGeometry> tgeo)
+      : m_guider(guider), m_tgeo(tgeo) {}
+
+  /// The whole purpose of this function is to set currentSurface, if possible
+  template <typename propagator_state_t, typename stepper_t>
+  void status(propagator_state_t &state, const stepper_t &stepper) const {
+    const auto &logger = state.options.logger;
+    ACTS_VERBOSE(">>>>>>>> STATUS <<<<<<<<<");
+
+    if (state.navigation.navigationBreak)
+      return;
+
+    // TODO Is initialization correct for all cases?
+    if (state.navigation.navSurfaces.empty()) {
+      initialize(state, stepper);
+      return;
+    }
+
+    // Navigator status always resets the current surface
+    state.navigation.currentSurface = nullptr;
+
+    // Establish the surface status
+    auto surfaceStatus = stepper.updateSurfaceStatus(
+        state.stepping, *state.navigation.navSurfaceIter->representation,
+        false);
+
+    if (surfaceStatus == Acts::Intersection3D::Status::onSurface) {
+      // Set the current surface
+      state.navigation.currentSurface =
+          state.navigation.navSurfaceIter->representation;
+      ACTS_VERBOSE("Current surface set to  "
+                   << state.navigation.currentSurface->geometryId());
+
+      // Release Stepsize
+      ACTS_VERBOSE("Release Stepsize");
+      stepper.releaseStepSize(state.stepping);
+
+      // Reset state
+      state.navigation.navSurfaces.clear();
+      state.navigation.navSurfaceIter = state.navigation.navSurfaces.end();
+
+    } else if (surfaceStatus == Acts::Intersection3D::Status::reachable) {
+      ACTS_VERBOSE("Next surface reachable at distance  "
+                   << stepper.outputStepSize(state.stepping));
+    } else {
+      ACTS_VERBOSE(
+          "Surface unreachable or missed, hopefully the target(...) call "
+          "fixes this");
+    }
+  }
+
+  template <typename propagator_state_t, typename stepper_t>
+  void target(propagator_state_t &state, const stepper_t &stepper) const {
+    const auto &logger = state.options.logger;
+    ACTS_VERBOSE(">>>>>>>> TARGET <<<<<<<<<");
+    auto &navstate = state.navigation;
+
+    // TODO Workaround since we don't have a currentVolume yet
+    // In the case of the trial&error surface provider this does not happen,
+    // since it stops if no surface is reachable
+    if (!navstate.worldVolume->inside(stepper.position(state.stepping))) {
+      navstate.navigationBreak = true;
+      navstate.currentVolume = nullptr;
+      ACTS_VERBOSE("not inside worldVolume, stop navigation");
+      return;
+    }
+
+    // Fetch new targets if there are no candidates in state
+    if (navstate.navSurfaceIter == navstate.navSurfaces.end()) {
+      // This means also we are currently on a surface
```

Comment:
Does this have to be the case? Can't we have missed the last surface?

Reply:
Yeah, I think you are right... For this case one would have to stop the navigation I think...